commands/build.go

```diff
@@ -434,7 +456,263 @@ func buildCmd(dockerCli command.Cli, rootOpts *rootOptions) *cobra.Command {
+					}, true) // new session for detaching process group and tty
+					if err != nil {
+						return err
+					}
+					if err := wait(); err != nil {
+						return err
+					}
+					c, err = newBuildxClientAndCheck(filepath.Join(serverRoot, "buildx.sock"), 10, time.Second)
+					if err != nil {
+						return fmt.Errorf("cannot connect to the buildx server: %w", err)
+					}
+				}
+				defer func() {
+					if err := c.Close(); err != nil {
+						logrus.Warnf("failed to close server connection %v", err)
+					}
+				}()
+
+				f := ioset.NewSingleForwarder()
+				pr, pw := io.Pipe()
+				f.SetWriter(pw, func() io.WriteCloser {
+					pw.Close() // propagate EOF
+					logrus.Debug("propagating stdin close")
+					return nil
+				})
+				f.SetReader(os.Stdin)
+				builderOptions := buildOption2builderOptions(options)
+				ref, err := c.Build(ctx, builderOptions, pr, os.Stdout, os.Stderr, options.progress)
+				if err != nil {
+					return fmt.Errorf("failed to build: %w", err) // TODO: allow invoke even on error
+				}
+				if err := pw.Close(); err != nil {
+					logrus.Debug("failed to close stdin pipe writer")
+				}
+				if err := pr.Close(); err != nil {
+					logrus.Debug("failed to close stdin pipe reader")
+				}
+
+				// post-build operations
+				ctrl := &buildxController{c, serverRoot}
+				if options.invoke != "" {
+					// NOTE: We close client here to reset the I/O. In monitor mode, we let it manage I/O
+					//       and we don't handle I/O anymore.
+					pr2, pw2 := io.Pipe()
+					f.SetWriter(pw2, func() io.WriteCloser {
+						pw2.Close() // propagate EOF
+						return nil
+					})
+					con := console.Current()
+					if err := con.SetRaw(); err != nil {
+						if err := c.Disconnect(ctx, ref); err != nil {
+							logrus.Warnf("disconnect error: %v", err)
+						}
+						return errors.Errorf("failed to configure terminal: %v", err)
+					}
+					err = monitor.RunMonitor(ctx, ref, builderOptions, invokeConfig, ctrl, options.progress, pr2, os.Stdout, os.Stderr)
+					con.Reset()
+					if err := pw2.Close(); err != nil {
+						logrus.Debug("failed to close monitor stdin pipe reader")
+					}
+					if err != nil {
+						logrus.Warnf("failed to run monitor: %v", err)
+					}
+				} else {
+					if err := c.Disconnect(ctx, ref); err != nil {
+						logrus.Warnf("disconnect error: %v", err)
+					}
+					// If "invoke" isn't specified, further inspection ins't provided. Finish the buildx server.
+					if err := ctrl.Kill(ctx); err != nil {
+						return err
+					}
+				}
+
+				return nil
+			case controlModeLaunch:
+				//
+				// step 2: child of buildx client for launching and detaching buildx server
+				//
+				logrus.Debugf("start launch with args %v", args)
+
+				// start buildx server via "serve"
+				next := cInfo
+				next.Mode = controlModeServe
+				wait, err := launch(context.TODO(), &next, false)
+				if err != nil {
+					return err
+				}
+				go wait()
+				return nil
+			case controlModeServe:
```

Comment:
If we can, I think it makes sense to actually have a separate internal sub-command for this, instead of trying to distinguish by env var - only the client-side parts should need to be in this function.